worker/changestream/eventmultiplexer/eventmultiplexer.go

```diff
@@ -0,0 +1,296 @@
+
+		case subOpt := <-e.subscriptionCh:
+			sub := subOpt.subscription
+
+			// Create a new subscription and assign a unique ID to it.
+			e.subscriptions[sub.id] = sub
+
+			// No options were supplied, just add it to the all bucket, so
+			// they'll be included in every dispatch.
+			if len(subOpt.opts) == 0 {
+				e.subscriptionsAll[sub.id] = struct{}{}
+				continue
+			}
+
+			// Register filters to route changes matching the subscription criteria to
+			// the newly created subscription.
+			for _, opt := range subOpt.opts {
+				namespace := opt.Namespace()
+				e.subscriptionsByNS[namespace] = append(e.subscriptionsByNS[namespace], &eventFilter{
+					subscriptionID: sub.id,
+					changeMask:     opt.ChangeMask(),
+					filter:         opt.Filter(),
+				})
+				sub.topics[namespace] = struct{}{}
+			}
+
+		case subscriptionID := <-e.unsubscriptionCh:
+			sub, found := e.subscriptions[subscriptionID]
+			if !found {
+				continue
+			}
+
+			for topic := range sub.topics {
+				var updatedFilters []*eventFilter
+				for _, filter := range e.subscriptionsByNS[topic] {
+					if filter.subscriptionID == subscriptionID {
+						continue
+					}
+					updatedFilters = append(updatedFilters, filter)
+				}
+				e.subscriptionsByNS[topic] = updatedFilters
+			}
+
+			delete(e.subscriptions, subscriptionID)
+			delete(e.subscriptionsAll, subscriptionID)
+
+			sub.close()
+		}
+	}
+}
+
+func (e *EventMultiplexer) gatherSubscriptions(ch changestream.ChangeEvent) []*subscription {
+	subs := make(map[uint64]*subscription)
+
+	for id := range e.subscriptionsAll {
+		subs[id] = e.subscriptions[id]
+	}
+
+	traceEnabled := e.logger.IsTraceEnabled()
+	for _, subOpt := range e.subscriptionsByNS[ch.Namespace()] {
+		if _, ok := subs[subOpt.subscriptionID]; ok {
+			continue
+		}
+
+		if (ch.Type() & subOpt.changeMask) == 0 {
+			continue
+		}
+
+		if !subOpt.filter(ch) {
+			if traceEnabled {
+				e.logger.Tracef("filtering out change: %v", ch)
+			}
+			continue
+		}
+
+		if traceEnabled {
+			e.logger.Tracef("dispatching change: %v", ch)
+		}
+
+		subs[subOpt.subscriptionID] = e.subscriptions[subOpt.subscriptionID]
+	}
+
+	// By collecting the subs within a map to ensure that a sub can be only
+	// called once, we actually gain random ordering. This prevents subscribers
+	// from depending on the order of dispatches.
+	results := make([]*subscription, 0, len(subs))
+	for _, sub := range subs {
+		results = append(results, sub)
+	}
+	return results
+}
+
+// dispatchSet fans out the subscription requests against a given term of changes.
+// Each subscription signals the change in a asynchronous fashion, allowing
+// a subscription to not block another change within a given term.
+func (e *EventMultiplexer) dispatchSet(changeSet map[*subscription]ChangeSet) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	grp, grpCtx := errgroup.WithContext(e.catacomb.Context(ctx))
+	grp.SetLimit(len(changeSet))
```

Comment:
we don't need to set a limit here since the errgroup by default will be unlimited. This just reduces the performance of the errgroup.Go call.